File: sdks/veil-py/examples/smoke.py
#!/usr/bin/env python3
"""Minimal end-to-end smoke test for the Veil Python bindings.

Build libveil first:

    cd ../../core
    CGO_ENABLED=1 go build -buildmode=c-shared \
        -o ../sdks/veil-py/libveil.so ./pkg/cgo

Then:

    python examples/smoke.py path/to/client.yaml
"""

from __future__ import annotations

import sys
import time

import veil


def main() -> int:
    if len(sys.argv) < 2:
        print(f"usage: {sys.argv[0]} <path/to/client.yaml-or-json>")
        return 2

    print("library version:", veil.library_version())

    cfg_text = open(sys.argv[1], "r", encoding="utf-8").read()

    with veil.Veil(cfg_text) as v:
        v.start(on_event=lambda e: print(
            f"[event] kind={e.typed()!r} transport={e.transport} "
            f"remote={e.remote} tx={e.bytes_tx} rx={e.bytes_rx} msg={e.message}"
        ))
        for _ in range(10):
            time.sleep(1)
            print("metrics:", v.metrics())
        v.stop()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: sdks/veil-rs/examples/smoke.rs
//! Minimal end-to-end smoke test for the Veil Rust bindings.
//!
//! Build the libveil shared library first:
//!
//!     cd ../../core
//!     CGO_ENABLED=1 go build -buildmode=c-shared \
//!         -o ../sdks/veil-rs/libveil.so ./pkg/cgo
//!
//! Then run:
//!
//!     LD_LIBRARY_PATH=. cargo run --example smoke -- client.yaml
//!
//! On macOS / Windows substitute libveil.dylib / veil.dll and the
//! corresponding loader path variable.

use std::{env, fs, sync::Arc, time::Duration};

fn main() {
    let path = env::args()
        .nth(1)
        .expect("usage: smoke <path/to/client.yaml-or-json>");
    let cfg = fs::read_to_string(&path).expect("read config");

    println!("libveil version: {:?}", veil::Veil::library_version());

    let v = veil::Veil::create(&cfg).expect("create");
    let cb: veil::EventHandler = Arc::new(|e: veil::Event| {
        println!(
            "[event] kind={:?} transport={} remote={} bytes_tx={} bytes_rx={} msg={}",
            e.typed(),
            e.transport,
            e.remote,
            e.bytes_tx,
            e.bytes_rx,
            e.message
        );
    });
    v.start(Some(cb)).expect("start");

    // Hold the instance for a short while so a curl through the
    // SOCKS5 proxy from another shell would observe it.
    std::thread::sleep(Duration::from_secs(15));

    println!("metrics: {:?}", v.metrics());
    v.stop().expect("stop");
    drop(v); // forces destroy + free
    println!("done");
}
